pyro/contrib/oed/eig.py

```diff
@@ -0,0 +1,63 @@
+import pyro
+from pyro import poutine
+from pyro.contrib.oed.search import Search
+from pyro.infer import EmpiricalMarginal, Importance, SVI, Trace_ELBO
+from pyro.contrib.autoguide import mean_field_guide_entropy
+
+
+def vi_ape(model, design, observation_labels, vi_parameters, is_parameters):
+    """Estimates the average posterior entropy (APE) loss function using
+    variational inference (VI).
+
+    The APE loss function estimated by this method is defined as
+
+        :math:`APE(d)=E_{Y\sim p(y|\theta, d)}[H(p(\theta|y, d))]`
+
+    This is related to expected information gain (EIG) by the equation
+
+        :math:`EIG(d)=H[p(\theta)]-APE(d)`
+
+    in particular, minimising the loss is equivalent to maximising EIG.
+
+    :param function model: A pyro model accepting `design` as only argument.
+    :param torch.Tensor design: Tensor representation of design
+    :param observation_labels list or str: A subset of the sample sites
+        present in `model`. These sites are regarded as future observations
+        and other sites are regarded as latent variables over which a
+        posterior is to be inferred.
+    :param dict vi_parameters: Variational inference parameters which should include:
+        `optim`: an instance of `pyro.Optim`, `guide`: a guide function
+        compatible with `model`, `num_steps`: the number of VI steps to make
+    :param dict is_parameters: Importance sampling parameters for the
+        marginal distribution of :math:`Y`. May include `num_samples`: the number
+        of samples to draw from the marginal.
+    :return: Loss function estimate
+    :rtype: `torch.Tensor`
+
+    """
+
+    if isinstance(observation_labels, str):
+        observation_labels = [observation_labels]
+
+    def posterior_entropy(y_dist, design):
+        # Important that y_dist is sampled *within* the function
+        y = pyro.sample("conditioning_y", y_dist)
+        y_dict = {label: y[i, ...] for i, label in enumerate(observation_labels)}
+        conditioned_model = pyro.condition(model, data=y_dict)
+        posterior = SVI(conditioned_model, vi_parameters["guide"],
+                        vi_parameters["optim"], loss=Trace_ELBO())
+        with pyro.poutine.block():
+            for _ in range(vi_parameters["num_steps"]):
+                posterior.step(design)
+        # Recover the entropy
+        return mean_field_guide_entropy(vi_parameters["guide"], design)
+
+    y_dist = EmpiricalMarginal(
+        Importance(model, num_samples=is_parameters.get("num_samples", None)).run(design),
```

Comment:
What about using `EmpiricalMarginal(Importance(...), sites=observation_labels).run(design)` to pick out the observations correctly?

Reply:
I tried writing

    y_dist = EmpiricalMarginal(
        Importance(model, num_samples=is_parameters.get("num_samples", None)),
        sites=observation_labels).run(design)

doesn't work:

    AttributeError: 'EmpiricalMarginal' object has no attribute 'run'

Reply:
I think I misunderstood what you were suggesting

Reply:
That's... what I already had?? (see line below)

Reply:
Oh yeah, you're right, totally misread it